lib/streamlit/runtime/fragment.py

```diff
@@ -0,0 +1,78 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from abc import abstractmethod
+from typing import Any, Callable
+
+from typing_extensions import Protocol
+
+Fragment = Callable[[], Any]
+
+
+class FragmentStorage(Protocol):
+    """A key-value store for Fragments. Used to implement the @st.experimental_fragment
+    decorator.
+
+    We intentionally define this as its own protocol despite how generic it appears to
```

Comment:
This means that something like a version with cloudpickle serialization would apply the serialization/deserialization within the `set` and `get` methods, or?

Reply:
Yeah, that's what I'm currently thinking, although we should have a good amount of flexibility on what these protocol definitions actually end up being -- we'll be able to freely change them even after `st.experimental_fragment` is released as it'll be quite some time before we officially document the protocols or write an implementation aside from `MemoryFragmentStorage`